### PR TITLE
[TOPIC-BLE-LLCP] bluetooth: controller: llcp: Terminate procedure

### PR DESCRIPTION
**Pull request by sjanc**
This implementes initial support for link termination procedure and
unit tests.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

### CODE REVIEW

File: tests/bluetooth/controller/ctrl_terminate/src/main.c [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+	/* Tx Queue should have no LL Control PDU */
+	lt_rx_q_is_empty(&conn);
+
+	/* Done */
+	event_done(&conn);
+
+	/* There should not be a host notifications */
+	ut_rx_q_is_empty();
+}
```
Comment (kruithofa) [inline tests/bluetooth/controller/ctrl_terminate/src/main.c:78]: Shouldn't we verify here that the state is now set to DISCONNECTED?

**Comment by kruithofa** [comment]
LGTM, just wandering if we should check that at the end of the procedure the state is indeed 'disconnected'

**Comment by sjanc** [comment]
Yeah, I'm now reworking those tests a bit

File: tests/bluetooth/controller/ctrl_terminate/src/main.c [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+void test_terminate_sla_loc(void)
+{
+	test_terminate_loc(BT_HCI_ROLE_SLAVE);
+}
+
+void test_main(void)
+{
+	ztest_test_suite(muc,
```
Comment (thoh-ot) [inline tests/bluetooth/controller/ctrl_terminate/src/main.c:137]: I would rename 'muc' to 'term'

File: tests/bluetooth/controller/ctrl_terminate/src/main.c [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+	uint8_t err;
+	struct node_tx *tx;
+	struct node_rx_pdu *ntf;
+
+	struct pdu_data_llctrl_terminate_ind local_terminate_ind = {
+		.error_code = 0x06,
+	};
+
+	/* Role */
+	test_set_role(&conn, BT_HCI_ROLE_MASTER);
```
Comment (thoh-ot) [inline tests/bluetooth/controller/ctrl_terminate/src/main.c:96]: Shouldn't 'role' have been used here?

File: subsys/bluetooth/controller/ll_sw/ull_llcp.h [hunk 4]
```diff
@@ -54,6 +54,10 @@ struct ull_cp_conn {
 			uint8_t min_used_chans;
 		} muc;
 
+		/* Use by ACL Termination Procedure */
+		struct {
+			uint8_t error_code;
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.h:59]: Does the error_code need to be a persistent field? Isn't it only a transient value, as the link will be disconnected so there is no value in remembering it.

**Review by thoh-ot — CHANGES_REQUESTED** [review]

File: tests/bluetooth/controller/ctrl_terminate/src/main.c (hunk 2 above)
Comment (sjanc) [reply tests/bluetooth/controller/ctrl_terminate/src/main.c:137]: copy-paste, will fix :-)

File: tests/bluetooth/controller/ctrl_terminate/src/main.c (hunk 3 above)
Comment (sjanc) [reply tests/bluetooth/controller/ctrl_terminate/src/main.c:96]: ups, yeap, will fix

File: subsys/bluetooth/controller/ll_sw/ull_llcp.h (hunk 4 above)
Comment (sjanc) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.h:59]: after disconnect is acknowledged we need to send event to host with proper disconnect reason (also for locally initiated disconnect)

**Review by thoh-ot — APPROVED** [review]

**Pull request merged**